// puppeteer-core/lib/esm/puppeteer/common/bidi/Page.js
/**
 * Copyright 2022 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Page_instances, _Page_connection, _Page_subscribedEvents, _Page_onLogEntryAdded, _Page_evaluate;
import { Page as PageBase } from '../../api/Page.js';
import { stringifyFunction } from '../../util/Function.js';
import { ConsoleMessage } from '../ConsoleMessage.js';
import { isString } from '../util.js';
import { JSHandle } from './JSHandle.js';
import { BidiSerializer } from './Serializer.js';
/**
 * @internal
 */
export class Page extends PageBase {
    constructor(connection, contextId) {
        super();
        _Page_instances.add(this);
        _Page_connection.set(this, void 0);
        _Page_subscribedEvents.set(this, [
            'log.entryAdded',
        ]);
        __classPrivateFieldSet(this, _Page_connection, connection, "f");
        this._contextId = contextId;
        // TODO: Investigate an implementation similar to CDPSession
        this.connection.send('session.subscribe', {
            events: __classPrivateFieldGet(this, _Page_subscribedEvents, "f"),
            contexts: [this._contextId],
        });
        this.connection.on('log.entryAdded', __classPrivateFieldGet(this, _Page_instances, "m", _Page_onLogEntryAdded).bind(this));
    }
    async close() {
        await __classPrivateFieldGet(this, _Page_connection, "f").send('browsingContext.close', {
            context: this._contextId,
        });
        this.connection.send('session.unsubscribe', {
            events: __classPrivateFieldGet(this, _Page_subscribedEvents, "f"),
            contexts: [this._contextId],
        });
        this.connection.off('log.entryAdded', __classPrivateFieldGet(this, _Page_instances, "m", _Page_onLogEntryAdded).bind(this));
    }
    get connection() {
        return __classPrivateFieldGet(this, _Page_connection, "f");
    }
    async evaluateHandle(pageFunction, ...args) {
        return __classPrivateFieldGet(this, _Page_instances, "m", _Page_evaluate).call(this, false, pageFunction, ...args);
    }
    async evaluate(pageFunction, ...args) {
        return __classPrivateFieldGet(this, _Page_instances, "m", _Page_evaluate).call(this, true, pageFunction, ...args);
    }
}
_Page_connection = new WeakMap(), _Page_subscribedEvents = new WeakMap(), _Page_instances = new WeakSet(), _Page_onLogEntryAdded = function _Page_onLogEntryAdded(event) {
    var _a;
    if (isConsoleLogEntry(event)) {
        const args = event.args.map(arg => {
            return getBidiHandle(this, arg);
        });
        const text = args
            .reduce((value, arg) => {
            const parsedValue = arg.isPrimitiveValue
                ? BidiSerializer.deserialize(arg.bidiObject())
                : arg.toString();
            return `${value} ${parsedValue}`;
        }, '')
            .slice(1);
        this.emit("console" /* PageEmittedEvents.Console */, new ConsoleMessage(event.method, text, args, getStackTraceLocations(event.stackTrace)));
    }
    else if (isJavaScriptLogEntry(event)) {
        this.emit("console" /* PageEmittedEvents.Console */, new ConsoleMessage(event.level, (_a = event.text) !== null && _a !== void 0 ? _a : '', [], getStackTraceLocations(event.stackTrace)));
    }
}, _Page_evaluate = async function _Page_evaluate(returnByValue, pageFunction, ...args) {
    let responsePromise;
    const resultOwnership = returnByValue ? 'none' : 'root';
    if (isString(pageFunction)) {
        responsePromise = __classPrivateFieldGet(this, _Page_connection, "f").send('script.evaluate', {
            expression: pageFunction,
            target: { context: this._contextId },
            resultOwnership,
            awaitPromise: true,
        });
    }
    else {
        responsePromise = __classPrivateFieldGet(this, _Page_connection, "f").send('script.callFunction', {
            functionDeclaration: stringifyFunction(pageFunction),
            arguments: await Promise.all(args.map(arg => {
                return BidiSerializer.serialize(arg, this);
            })),
            target: { context: this._contextId },
            resultOwnership,
            awaitPromise: true,
        });
    }
    const { result } = await responsePromise;
    if ('type' in result && result.type === 'exception') {
        throw new Error(result.exceptionDetails.text);
    }
    return returnByValue
        ? BidiSerializer.deserialize(result.result)
        : getBidiHandle(this, result.result);
};
/**
 * @internal
 */
export function getBidiHandle(context, result) {
    if ((result.type === 'node' || result.type === 'window') &&
        context._contextId) {
        // TODO: Implement ElementHandle
        return new JSHandle(context, result);
    }
    return new JSHandle(context, result);
}
function isConsoleLogEntry(event) {
    return event.type === 'console';
}
function isJavaScriptLogEntry(event) {
    return event.type === 'javascript';
}
function getStackTraceLocations(stackTrace) {
    const stackTraceLocations = [];
    if (stackTrace) {
        for (const callFrame of stackTrace.callFrames) {
            stackTraceLocations.push({
                url: callFrame.url,
                lineNumber: callFrame.lineNumber,
                columnNumber: callFrame.columnNumber,
            });
        }
    }
    return stackTraceLocations;
}
